landoapi/workers/revision_worker.py

```diff
@@ -0,0 +1,428 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from __future__ import annotations
+
+import io
+import logging
+from pathlib import Path
+from itertools import chain
+
+import networkx as nx
+from mots.config import FileConfig
+from mots.directory import Directory, QueryResult
+
+from landoapi.hg import HgRepo
+from landoapi.models.revisions import Revision
+from landoapi.models.revisions import RevisionStatus as RS
+from landoapi.phabricator import get_conduit_data
+from landoapi.repos import repo_clone_subsystem
+from landoapi.storage import db, _lock_table_for
+from landoapi.workers.base import RevisionWorker
+
+logger = logging.getLogger(__name__)
+
+
+DIFF_CONTEXT_SIZE = 5000
+
+
+class StackGraph(nx.DiGraph):
+    def __eq__(self, G):
+        return nx.utils.misc.graphs_equal(self, G)
+
+    @property
+    def revisions(self):
+        return self.nodes
+
+
+def get_active_repos(repo_config: dict) -> list[str]:
+    """Query Phabricator to determine PHIDs of active repos."""
+    repos = [repo for repo in repo_config if repo.use_revision_worker]
+    repo_phids = get_conduit_data(
+        "diffusion.repository.search",
+        constraints={"shortNames": [r.short_name for r in repos]},
+    )
+    return [r["phid"] for r in repo_phids]
+
+
+def get_stacks(revisions: dict[str, dict]) -> list:
+    """Returns a stack with revision PHIDs as nodes.
+
+    This method fetches unique stacks from a list of stack graphs. This
+    is because Phabricator returns different forms of the same stack graph
+    in each revision.
+
+    This method will return a list of StackGraph objects.
+    """
+    stacks = [r["fields"]["stackGraph"] for r in revisions.values()]
+    parsed = [StackGraph(s).reverse() for s in stacks]
+
+    filtered = []
+    for stack in parsed:
+        if stack not in filtered:
+            filtered.append(stack)
+    return filtered
+
+
+def get_phab_revisions(statuses: list[str] | None = None) -> dict[int, dict]:
+    """Get a list of revisions of given statuses."""
+    statuses = statuses or [
+        "draft",
+        "needs-review",
+        "accepted",
+        "published",
+        "changes-planned",
+    ]
```

Comment:
It limits how many revisions we are synchronizing based on those that we need. The list will probably be tweaked in the future (e.g., maybe we don't sync drafts, or changes-planned, since those won't be landable anyway.)